File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    bool validate(TreeNode* root, int lft=INT_MIN, int rgt=INT_MAX) {
        if(root==nullptr)  return true;
        if(root->val<=lft || root->val>=rgt) return false;

        return validate(root->left, lft, root->val) &&
            validate(root->right, root->val, rgt);
    }
public:
    bool isValidBST(TreeNode* root) {
        return validate(root);
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root==null) {
            return root;
        }

        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);

        if ((left!=null && right!=null && left.val==p.val && right.val==q.val)
            || (left!=null && right!=null && left.val==q.val && right.val==p.val)
            || (left!=null && root.val==p.val && left.val==q.val)
            || (left!=null && root.val==q.val && left.val==p.val)
            || (right!=null && root.val==p.val && right.val==q.val)
            || (right!=null && root.val==q.val && right.val==p.val)) { return root; }

        if (left!=null && (left.val==p.val || left.val==q.val)) { return left; }
        if (right!=null && (right.val==p.val || right.val==q.val)) { return right; }
        if (root.val==p.val || root.val==q.val) return root;
        if(left!=null) return left;
        if(right!=null) return right;

        return null;
    }
}
